### ISSUE TITLE
Failed to show text document error using FileSystemProvider

### ISSUE THREAD

**Issue by dnicolson** [microsoft/vscode#84046]
- VSCode Version: 1.39.2
- OS Version: 10.14.6 (18G103)

Steps to Reproduce:

1. Create an extension with the default options
2. Add the code below
3. Run the extension
4. Open a file ending with "test"

A new text window correctly opens, however there is an error message:
```
rejected promise not handled within 1 second: Error: Failed to show text document myprovider:/Users/dnicolson/test, should show in editor #vs.editor.ICodeEditor:1,$model2
extensionHostProcess.js:787
stack trace: Error: Failed to show text document myprovider:/Users/dnicolson/test, should show in editor #vs.editor.ICodeEditor:1,$model2
	at t.ExtHostEditors.showTextDocument._proxy.$tryShowTextDocument.then.t (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:615:78)
```

<details><summary>extension.ts</summary>

```typescript

import * as vscode from 'vscode';
const util = require('util');

class MyFileSystemProvider implements vscode.FileSystemProvider {
	readFile(uri: vscode.Uri): Uint8Array {
		return new util.TextEncoder('utf-8').encode('Text content');
	}

	writeFile(uri: vscode.Uri, content: Uint8Array, options: { create: boolean; overwrite: boolean; }): void {}

	stat(uri: vscode.Uri): vscode.FileStat | Thenable<vscode.FileStat> {
		return ({ type: vscode.FileType.File }) as vscode.FileStat;
	}

	onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> = new vscode.EventEmitter<vscode.FileChangeEvent[]>().event;
	createDirectory(uri: vscode.Uri) { }
	delete(uri: vscode.Uri, options: { recursive: boolean; }) { }
	readDirectory(uri: vscode.Uri) { return []; }
	rename(oldUri: vscode.Uri, newUri: vscode.Uri, options: { overwrite: boolean; }) { }
	watch(uri: vscode.Uri, options: { recursive: boolean; excludes: string[]; }) { return new vscode.Disposable(() => { }); }
}

export function activate(context: vscode.ExtensionContext) {
	vscode.workspace.registerFileSystemProvider('myprovider', new MyFileSystemProvider(), {
		isCaseSensitive: process.platform === 'linux'
	});

	vscode.workspace.onDidOpenTextDocument(async document => {
		if (document.fileName.match(/test$/)) {
			const uri = vscode.Uri.file(document.fileName).with({scheme: 'myprovider'});
      const doc = await vscode.workspace.openTextDocument(uri);
      await vscode.window.showTextDocument(doc, { preview: false });
		}
	});
}

export function deactivate() {}
```
</details>

<details><summary>package.json</summary>

```typescript
"activationEvents": [
	"onFileSystem:myprovider",
	"*"
]
```
</details>

Does this issue occur when all extensions are disabled?: Yes

**Comment by jrieken** [comment]
Please try this again with a complete implementation of `stat` - the one showing above only sets the type and I believe we also use size and the times.

**Comment by dnicolson** [comment]
The error is still displayed after including the contents of `fs.lstatSync(uri.path)`.

**Comment by jrieken** [comment]
> The error is still displayed after including the contents of fs.lstatSync(uri.path)

Can sure what that means. Can you update your sample with your changes?

**Comment by dnicolson** [comment]
I haven't updated the original but here are the complete methods of what I have attempted:

```
stat(uri: vscode.Uri): vscode.FileStat | Thenable<vscode.FileStat> {
    const stat = fs.lstatSync(uri.path);
    return ({ type: vscode.FileType.File, ...stat, name: 'test' }) as vscode.FileStat;
}
```
```
stat(uri: vscode.Uri): vscode.FileStat | Thenable<vscode.FileStat> {
    return ({ type: vscode.FileType.File, ctime: Date.now(), mtime: Date.now(), size: 0, name: 'test' }) as vscode.FileStat;
}
```

**Comment by jrieken** [comment]
can you try with our sample provider: https://github.com/microsoft/vscode-extension-samples/blob/master/fsprovider-sample/src/fileSystemProvider.ts

**Comment by dnicolson** [comment]
I used the `MemFS` class and the same problem exists.

Files can be read without error but after adding `.with({scheme: 'myprovider'});` the `stat` and `readFile` methods must be replaced.

Everything works despite the error.

**Comment by jrieken** [comment]
> I used the MemFS class and the same problem exists.

Files can be read without error but after adding .with({scheme: 'myprovider'}); the stat and readFile methods must be replaced.

Sorry, I don't understand what that means

**Comment by dnicolson** [comment]
I used all the methods from `MemFS`, the sample provider linked above.

The class works without error, but the scheme needs to be changed to create a virtual file system (`myprovider:`). After `stat` and `readFile` methods are updated the code works (with the aforementioned error).

**Comment by jrieken** [comment]
I have tried your snippet and it works for me. I had to remove `return new util.TextEncoder('utf-8').encode('Text content');` because I don't have the `util`-lib. Instead I have used Buffer.fromString('Text content') and when opening a test file I see the fake-file being opened.

**Comment by dnicolson** [comment]
It looks like the issue is fixed with the newly released version 1.40.
